### PR TITLE
[Services] Restart database service upon unexpected critical process exit.

### PR DESCRIPTION
**Pull request by yozhao101**
Signed-off-by: Yong Zhao <yozhao@microsoft.com>

- What I did
Restart database service if one of critical processes running in database container exited or crashed abnormally.

- How I did it
Generally I follow the framework created by Joe to implement this feature in database container.
**First**, add supervisor-proc-exit-listener event listener option in Supervisord configuration file in database docker container. Supervisord will read a list of critical processes for which to monitor the unexpected crashed and exited.
**Second**, configure database.service to always auto-restart the service if it stops, with a delay of 30 seconds. Also set a rate limit of 3 restarts within 20 minutes (1200 seconds).

- How to verify it
On your switch device, please use **docker ps** command to list all running docker containers.
Then use **docker exec -it container_id bash** to login target container. Typing **top** command
on the shell will display all the processes dynamically and you will spot the process id of one
of the critical processes. Finally type the command **kill -9 process_id** to terminate one process.
After exiting the container, you can use **watch -n 1 docker ps** to dynamically see the restart
of database container.

### CODE REVIEW

File: files/build_templates/dhcp_relay.service.j2 [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [Unit]
 Description=DHCP relay container
-Requires=updategraph.service
+Requires=updategraph.service database.service
```
Comment (jleveque) [inline files/build_templates/dhcp_relay.service.j2]: Is it necessary to explicitly add database.service here? This requires updategraph.service, which in turn requires config-setup.service which requires database.service. Did you find this chain of dependencies was not sufficient?

File: files/scripts/supervisor-proc-exit-listener [hunk 2]
```diff
@@ -52,24 +52,25 @@ def main(argv):
...
+                    sys.exit(3)
+
+                restart_feature = container_features_table[container_name].get('auto_restart')
+                if not restart_feature:
+                    syslog.syslog(syslog.LOG_ERR, "Unable to determine auto-restart feature status for container '{}'. Exiting...".format(container_name))
+                    sys.exit(4)
 
             # If auto-restart feature is enabled and a critical process exited unexpectedly, terminate supervisor
-            if restart_feature == 'enabled' and expected == 0 and (processname in critical_processes or groupname in critical_processes):
+            if (container_name == 'database' or restart_feature == 'enabled') and expected == 0 and (processname in critical_processes or groupname in critical_processes):
```
Comment (jleveque) [inline files/scripts/supervisor-proc-exit-listener]: This line has gotten long and complex. Can you either break it into two lines or two conditionals?

Comment (yozhao101) [reply files/scripts/supervisor-proc-exit-listener]: Will fix it.

File: files/build_templates/dhcp_relay.service.j2 (hunk 1 above)
Comment (yozhao101) [reply files/build_templates/dhcp_relay.service.j2]: I will look into this.

Comment (yozhao101) [reply files/build_templates/dhcp_relay.service.j2]: I checked the service files and the current chain of dependencies is already sufficient.

Comment (jleveque) [reply files/build_templates/dhcp_relay.service.j2]: Did you test that restarting the database container actually causes the entire chain of dependent services to restart?

File: files/scripts/supervisor-proc-exit-listener [hunk 3]
```diff
@@ -52,24 +52,28 @@ def main(argv):
...
+
+                restart_feature = container_features_table[container_name].get('auto_restart')
+                if not restart_feature:
+                    syslog.syslog(syslog.LOG_ERR, "Unable to determine auto-restart feature status for container '{}'. Exiting...".format(container_name))
+                    sys.exit(4)
+
+            # If container is database or auto-restart feature is enabled and at the same time 
+            # a critical process exited unexpectedly, terminate supervisor
+            if (container_name == 'database' or restart_feature == 'enabled') and expected == 0 and 
+                    (processname in critical_processes or groupname in critical_processes):
```
Comment (jleveque) [inline files/scripts/supervisor-proc-exit-listener]: Is this tested? Does it work as-is, or do you need to add parentheses around the entire conditional due to the line break?

Comment (yozhao101) [reply files/scripts/supervisor-proc-exit-listener]: Added a pair of parentheses.

File: files/build_templates/dhcp_relay.service.j2 (hunk 1 above)
Comment (yozhao101) [reply files/build_templates/dhcp_relay.service.j2]: Actually I tested it today but I want to test it again.

Comment (yozhao101) [reply files/build_templates/dhcp_relay.service.j2]: Checked again and it did work.

**Pull request merged**